From e1d1df5c09e79a553fc048400801609a5d246d3c Mon Sep 17 00:00:00 2001
From: Scott Wood <scottwood@freescale.com>
Date: Thu, 19 Nov 2009 06:42:00 +0800
Subject: [PATCH 028/233] powerpc/e500mc: Add an idle loop based on the wait instruction

Extracted from vendor drop QorIQ-DPAA-SDK-V1-20110609-systembuilder.iso
0169-powerpc-e500mc-Add-an-idle-loop-based-on-the-wait-in.patch

e500mc cannot doze due to an erratum, but it has a "wait" instruction
that is similar.  It is implemented as a separate idle loop rather than
an inner-loop power_save() callback, because we don't want IRQs or
IRQ-off tracing disabled, and most of the rest of cpu_idle() is
unnecessary even if possibly harmless.

Note that on platforms that select this idle loop will not be able to use
nap-on-idle.

The previous wait idle loop is generalized into a "simple" loop for
power_save callbacks that don't poll and are entered with interrupts
enabled.

Signed-off-by: Scott Wood <scottwood@freescale.com>
---
 arch/powerpc/include/asm/machdep.h     |    2 ++
 arch/powerpc/kernel/idle.c             |   27 +++++++++++++++++++++++++++
 arch/powerpc/platforms/85xx/p4080_ds.c |    4 +++-
 3 files changed, 32 insertions(+), 1 deletions(-)

diff --git a/arch/powerpc/include/asm/machdep.h b/arch/powerpc/include/asm/machdep.h
index 579b9fb..c2d285a 100644
--- a/arch/powerpc/include/asm/machdep.h
+++ b/arch/powerpc/include/asm/machdep.h
@@ -293,6 +293,8 @@ extern void e500_idle(void);
 extern void power4_idle(void);
 extern void power4_cpu_offline_powersave(void);
 extern void ppc6xx_idle(void);
+extern void cpu_idle_simple(void);
+extern void ppc_wait(void);
 
 /*
  * ppc_md contains a copy of the machine description structure for the
diff --git a/arch/powerpc/kernel/idle.c b/arch/powerpc/kernel/idle.c
index 049dda6..18abe7b 100644
--- a/arch/powerpc/kernel/idle.c
+++ b/arch/powerpc/kernel/idle.c
@@ -13,6 +13,8 @@
  *
  * 32-bit and 64-bit versions merged by Paul Mackerras <paulus@samba.org>
  *
+ * Portions Copyright 2009-2010 Freescale Semiconductor, Inc.
+ *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License
  * as published by the Free Software Foundation; either version
@@ -32,6 +34,7 @@
 #include <asm/time.h>
 #include <asm/machdep.h>
 #include <asm/smp.h>
+#include <asm/ppc-opcode.h>
 
 #ifdef CONFIG_HOTPLUG_CPU
 #define cpu_should_die()	cpu_is_offline(smp_processor_id())
@@ -102,6 +105,30 @@ void cpu_idle(void)
 	}
 }
 
+void cpu_idle_simple(void)
+{
+	while (1) {
+		tick_nohz_stop_sched_tick(1);
+
+		while (!need_resched() && !cpu_should_die())
+			ppc_md.power_save();
+
+		tick_nohz_restart_sched_tick();
+
+		if (cpu_should_die())
+			cpu_die();
+
+		preempt_enable_no_resched();
+		schedule();
+		preempt_disable();
+	}
+}
+
+void ppc_wait(void)
+{
+	asm volatile(PPC_WAIT(0));
+}
+
 int powersave_nap;
 
 #ifdef CONFIG_SYSCTL
diff --git a/arch/powerpc/platforms/85xx/p4080_ds.c b/arch/powerpc/platforms/85xx/p4080_ds.c
index 8417046..134c724 100644
--- a/arch/powerpc/platforms/85xx/p4080_ds.c
+++ b/arch/powerpc/platforms/85xx/p4080_ds.c
@@ -3,7 +3,7 @@
  *
  * Maintained by Kumar Gala (see MAINTAINERS for contact information)
  *
- * Copyright 2009 Freescale Semiconductor Inc.
+ * Copyright 2009-2010 Freescale Semiconductor Inc.
  *
  * This program is free software; you can redistribute  it and/or modify it
  * under  the terms of  the GNU General  Public License as published by the
@@ -68,6 +68,8 @@ define_machine(p4080_ds) {
 	.restart		= fsl_rstcr_restart,
 	.calibrate_decr		= generic_calibrate_decr,
 	.progress		= udbg_progress,
+	.idle_loop		= cpu_idle_simple,
+	.power_save		= ppc_wait,
 };
 
 machine_device_initcall(p4080_ds, corenet_ds_publish_devices);
-- 
1.7.0.4

